_posts/2016-04-17-sorted-pagination-in-cassandra.md

```diff
@@ -0,0 +1,178 @@
+author: Felipe Fernández
+image:
+    src: /assets/img/custom/blog/ring.png
+tags:
+- distributed systems
+- cassandra
+- data modeling
+
+---
+
+Cassandra is a fantastic database for different use cases. There are different situations when you need to twist Cassandra a little and studying one of those could be a helpful exercise to understand better what is Cassandra about. Databases are complex beasts, approaching them with the right level of abstraction is vital. Their final goal is not storing data per se, but make that data accessible. Those read patterns will define which database is the best tool for the job.
+
+## Time series in Cassandra
+
+A time serie is a collection of data related to some variable. Facebook's timeline would be a great example. A user will write a serie of posts over time. The access patterns to that data will be something like 'return the 20 last posts of user 1234'. The DDL of a table that models that query would be:
+
+```
+CREATE TABLE timeline (
+	user_id uuid,
+	post_id timeuuid,
+	content text,
+	PRIMARY KEY (user_id, post_id)
+)
+WITH CLUSTERING ORDER BY (post_id DESC);
+```
+
+In Cassandra Primary Keys are formed by Partition Keys and Clustering Keys. Primary keys enforce the uniqueness of some cells in a different way of relational database. There is no strong enforcement of that uniqueness, if you try to insert some cell related to an already existing primary key, that will be updated. Also the other way around: a 'missing' update will end up as insert. That's called upsert.
+
+Partition keys ensure in which node of the cluster the data is going to live. If you include at least one clustering key, the partition key will identify N rows. That could be confusing for someone coming from traditional relational databases. Cassandra does its best trying to bring its concepts into SQL terminology, but sometimes it could be weird for newbies. An example of Timeline table would be:
+
+```
+user_id--------------------------------post_id--------content
+346e896a-c6b4-4d4e-826d-a5a9eda50636---today----------Hi
+346e896a-c6b4-4d4e-826d-a5a9eda50636---yesterday------Hola
+346e896a-c6b4-4d4e-826d-a5a9eda50636---one week ago---Bye
+346e896a-c6b4-4d4e-826d-a5a9eda50636---two weeks ago--Ciao
+```
+
+In order to understand the example I converted post_id values into something that makes sense for the reader. As you can see there are several values with the same partition key (user_id) and that works as we defined a clustering key (post_id) that clusters those values and sort them (descendantly in this case). Remember that uniqueness is defined by the primary key (partition plus clustering key) so if we insert a row identified with '346e896a-c6b4-4d4e-826d-a5a9eda50636' and 'today' the content will be updated. Nothing gets really updated in disk as Cassandra works with immutable structures in disk, but at read time different writes with the same primary key will be resolved by 'last-timestamp-wins'.
```

Comment:
I think you'll have to go the long way and say "in descending order"
On Apr 18, 2016 8:47 AM, "Robert Taylor" <notifications@github.com> wrote:

> In _posts/2016-04-17-sorted-pagination-in-cassandra.md
> <https://github.com/codurance/site/pull/155#discussion_r60016786>:
>
> > +WITH CLUSTERING ORDER BY (post_id DESC);
> > +```
> > +
> > +In Cassandra Primary Keys are formed by Partition Keys and Clustering Keys. Primary keys enforce the uniqueness of some cells in a different way of relational database. There is no strong enforcement of that uniqueness, if you try to insert some cell related to an already existing primary key, that will be updated. Also the other way around: a 'missing' update will end up as insert. That's called upsert.
> > +
> > +Partition keys ensure in which node of the cluster the data is going to live. If you include at least one clustering key, the partition key will identify N rows. That could be confusing for someone coming from traditional relational databases. Cassandra does its best trying to bring its concepts into SQL terminology, but sometimes it could be weird for newbies. An example of Timeline table would be:
> > +
> > +```
> > +user_id--------------------------------post_id--------content
> > +346e896a-c6b4-4d4e-826d-a5a9eda50636---today----------Hi
> > +346e896a-c6b4-4d4e-826d-a5a9eda50636---yesterday------Hola
> > +346e896a-c6b4-4d4e-826d-a5a9eda50636---one week ago---Bye
> > +346e896a-c6b4-4d4e-826d-a5a9eda50636---two weeks ago--Ciao
> > +```
> > +
> > +In order to understand the example I converted post_id values into something that makes sense for the reader. As you can see there are several values with the same partition key (user_id) and that works as we defined a clustering key (post_id) that clusters those values and sort them (descendantly in this case). Remember that uniqueness is defined by the primary key (partition plus clustering key) so if we insert a row identified with '346e896a-c6b4-4d4e-826d-a5a9eda50636' and 'today' the content will be updated. Nothing gets really updated in disk as Cassandra works with immutable structures in disk, but at read time different writes with the same primary key will be resolved by 'last-timestamp-wins'.
>
> '...sort them'
> ->
> '...sorts them'.
>
> Also don't think there is such a word as 'descendantly'. How about just
> 'descending'?
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/codurance/site/pull/155/files/e481812bf3253ba58fd2cccfc393eff6ab76863c#r60016786>
>